### PR TITLE
.Net: Use IReadOnlyDictionary for metadata dictionaries

### PR DESCRIPTION
**Pull request by stephentoub**
And then stop making copies each time we hand one out.

Fixes https://github.com/microsoft/semantic-kernel/issues/4233

### CODE REVIEW

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs [hunk 1]
```diff
@@ -86,4 +83,41 @@ public IReadOnlyList<OpenAIFunctionToolCall> GetOpenAIFunctionToolCalls()
 
         return Array.Empty<OpenAIFunctionToolCall>();
     }
+
+    private static IReadOnlyDictionary<string, object?>? CreateMetadataDictionary(
+        IReadOnlyList<ChatCompletionsToolCall> toolCalls,
+        IReadOnlyDictionary<string, object?>? original)
+    {
+        // We only need to augment the metadata if there are any tool calls.
+        if (toolCalls?.Count > 0)
```
Comment (SergeyMenshykh) [inline dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs:92]: The toolCalls parameter is not nullable, right?

**Review by SergeyMenshykh — APPROVED** [review]

**Review by RogerBarreto — APPROVED** [review]

**Comment by RogerBarreto** [comment]
Thanks for the change Stephen!

File: dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs (hunk 1 above)
Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs:92]: It is an `Optional<IReadOnlyList>`, not sure how this surface after the deserialization.

https://github.com/Azure/azure-sdk-for-net/blob/9a3cbe25fda60daeb7539707ad8d58dd19aa700a/sdk/openai/Azure.AI.OpenAI/src/Generated/ChatResponseMessage.Serialization.cs#L25C13-L25C61

Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.OpenAI/AzureSdk/OpenAIChatMessageContent.cs]: Addressed

**Review by dmytrostruk — APPROVED** [review]

**Pull request merged**